x-pack/plugins/apm/public/components/shared/Links/url_helpers.ts

```diff
@@ -16,7 +16,19 @@ export function toQuery(search?: string): APMQueryParamsRaw {
 }
 
 export function fromQuery(query: APMQueryParams) {
-  return qs.stringify(query);
+  // we have to avoid encoding range params because they cause
+  // Kibana angular to decode them and append them to the existing
+  // URL as an encoded hash /shrug
+  const encoded = mapValues(query, (value, key) => {
+    if (['rangeFrom', 'rangeTo'].includes(key!)) {
+      return value;
+    }
+    return qs.escape(value.toString());
+  });
+
+  return qs.stringify(encoded, '&', '=', {
+    encodeURIComponent: (value: string) => value
+  });
```

Comment:
I'm not even sure if I like the follow proposal but at least it's super explicit (which I do like):

```ts
export function fromQuery2(query: QueryParams) {
  const {
    detailTab,
    flyoutDetailTab,
    kuery,
    page,
    sortDirection,
    sortField,
    spanId,
    traceId,
    transactionId,
    waterfallItemId
  } = query;

  const search = [];

  if (detailTab) {
    search.push(`detailTab=${detailTab}`);
  }

  if (flyoutDetailTab) {
    search.push(`flyoutDetailTab=${flyoutDetailTab}`);
  }

  if (kuery) {
    search.push(`kuery=${encodeURIComponent(kuery)}`);
  }

  // ...

  return search.join('&');
}
```

This way it's very clear how each param is encoded.